File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
trees = list(map(int, input().split()))
trees.sort(reverse=True)

def get_home_trees(height):
    home = 0
    for tree in trees:
        if tree <= height:
            break
        home += tree - height
        
    return home

answer = 0
def binary_search(start, end):
    global answer

    height = (start + end) // 2
    home = get_home_trees(height)
    # print(height, home)
    
    if home == m:
        answer = height
        return 
    if start == end or height == end or height == start:
        answer = height
        return
    
    if home < m:
        binary_search(start, height)
    if home > m:
        binary_search(height, end)

binary_search(0, max(trees))

print(answer)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const [N, M] = input[0].split(' ').map(Number)
const trees = input[1].split(' ').map(Number)
trees.sort((a, b) => a - b)

let answer = 0
let left = 0
let right = Math.max(...trees)
while(left <= right) {
  const mid = Math.floor((left + right) / 2)
  const remain = cut_trees(mid)

  if (M <= remain) {
    answer = Math.max(answer, mid)
    left = mid + 1
  } else {
    right = mid - 1
  }
}
console.log(answer)

function cut_trees(max_height) {
  let remain = 0
  for (const tree of trees) {
    remain += Math.max(0, tree - max_height)
  }

  return remain
}